spock-specs/src/test/groovy/org/spockframework/smoke/mock/MultiThreaded.groovy

```diff
@@ -2,33 +2,60 @@ package org.spockframework.smoke.mock
 
 import spock.lang.Issue
 import spock.lang.Specification
+import spock.lang.Timeout
 
 import java.util.concurrent.CountDownLatch
-import java.util.concurrent.Executors
-import java.util.concurrent.TimeUnit
 
-class MultiThreaded extends Specification {
+import static java.util.concurrent.TimeUnit.SECONDS
 
+class MultiThreaded extends Specification {
   @Issue("https://github.com/spockframework/spock/issues/583")
   def "mocks can block and wait on conditions"() {
     given:
-    CountDownLatch latch = new CountDownLatch(2)
-    List aList = Mock ()
-    def exec = Executors.newFixedThreadPool(2)
+    CountDownLatch latch1 = new CountDownLatch(1)
+    CountDownLatch latch2 = new CountDownLatch(1)
+    CountDownLatch latch3 = new CountDownLatch(2)
+    List aList = Mock()
 
     when:
-    exec.submit { aList.get(0) }
-    exec.submit { aList.get(0) }
-    exec.shutdown()
-    latch.await()
+    Thread.start { aList.get(0); latch3.countDown() }
+    Thread.start { aList.get(0); latch3.countDown() }
+    assert latch1.await(30, SECONDS)
+    assert latch2.await(30, SECONDS)
 
     then:
-    exec.awaitTermination(30, TimeUnit.SECONDS)
+    latch3.await(10, SECONDS)
+
+    1 * aList.get(_) >> {
+      latch1.countDown()
+      assert latch2.await(10, SECONDS)
+    }
+
+    1 * aList.get(_) >> {
+      latch2.countDown()
+      assert latch1.await(10, SECONDS)
+    }
```

Comment:
Sure, why not, good idea 👌

Reply:
Too good idea. :-( This fails sometimes. I had it fail on 1. try, on 513. try, and on 82,643. try. :-/

I introduced another race condition when `SingleResponseGenerator.respond` is called outside of synchronization (as that is what would deadlock) but due to that now does set the `endOfCycle` too late, when before already the next interaction is requested from the response chain, so again the first response closure is returned, which then fails my improved test.

But if I do the `respond` call within the synchronization, response chain, then the two closures would again deadlock as long as they are in the same response chain.

Precalculating the value to be deliivered in the synchronization and then just returning it outside would not change anything, as the precalculating still means calling the closure.

Separating the `endOfCycle` modification out from `respond` might not work properly, as for the iterable response generator you can only ask the iterator for `hasNext` and that does not change unless you called `next()`. But calling `next` again is something that should not happen inside the synchronization as it could deadlock again.

🙈

Reply:
An alternative might be to just not support this within one response chain, but it would be nicer if it worked.
Because especially if an iterator is involved, you might not be able to split it into two iterations.

Reply:
Ok, having slept over it and brooded over it while working, I think I now properly solved it by adding yet another layer of indirection. :-D

At least the tests stayed green now the first 150_000 tries each. :-D